packages/service-library/src/servicelib/docker_utils.py

```diff
@@ -132,6 +132,68 @@ class _PulledStatus:
     extracted: int = 0
 
 
+async def _parse_pull_information(
+    parsed_progress: _DockerPullImage, *, layer_id_to_size: dict[str, _PulledStatus]
+):
+    match parsed_progress.status.lower():
+        case progress_status if any(
+            msg in progress_status
+            for msg in [
+                "pulling from",
+                "pulling fs layer",
+                "waiting",
+                "digest: ",
+            ]
+        ):
+            # nothing to do here
+            pass
+        case "downloading":
+            assert parsed_progress.id  # nosec
+            assert parsed_progress.progress_detail  # nosec
+            assert parsed_progress.progress_detail.current  # nosec
+
+            layer_id_to_size.setdefault(
+                parsed_progress.id,
+                _PulledStatus(parsed_progress.progress_detail.total or 0),
+            ).downloaded = parsed_progress.progress_detail.current
+        case "verifying checksum" | "download complete":
+            assert parsed_progress.id  # nosec
+            layer_id_to_size.setdefault(
+                parsed_progress.id, _PulledStatus(0)
+            ).downloaded = layer_id_to_size.setdefault(
+                parsed_progress.id, _PulledStatus(0)
+            ).size
+        case "extracting":
+            assert parsed_progress.id  # nosec
+            assert parsed_progress.progress_detail  # nosec
+            assert parsed_progress.progress_detail.current  # nosec
+            layer_id_to_size.setdefault(
+                parsed_progress.id,
+                _PulledStatus(parsed_progress.progress_detail.total or 0),
+            ).extracted = parsed_progress.progress_detail.current
+        case "pull complete":
+            assert parsed_progress.id  # nosec
+            layer_id_to_size.setdefault(
+                parsed_progress.id, _PulledStatus(0)
+            ).extracted = layer_id_to_size[parsed_progress.id].size
+        case progress_status if any(
+            msg in progress_status
+            for msg in [
```

Comment:
as you already know, the problem I of this approach is that it will fail again as soon as docker changes their logging texts :-(
Have you consider using fuzzy matching for the strings? Python has a builtin library `difflib` that can be used to check the similarity between strings. SEE [this example](https://docs.python.org/3/library/difflib.html#sequencematcher-examples)

Chatty also wrote a small example for this use case and pointed out that https://github.com/seatgeek/thefuzz?tab=readme-ov-file#usage is another popular library

```python
import difflib

# Sample log data (this would typically come from a file)
log_data = """
Initialization complete.
Starting data processing...
Data processing is 50% done.
Data processing finished successfully.
"""

# Keywords to search for
keywords = ["initialization complete", "starting data processing", "data processing finished"]

# Function to find fuzzy matches in the log
def find_fuzzy_matches(log, keywords, threshold=0.8):
    matches = []
    log_lines = log.split('\n')
    for keyword in keywords:
        for line in log_lines:
            # Calculate the similarity ratio
            match_ratio = difflib.SequenceMatcher(None, keyword, line).ratio()
            if match_ratio >= threshold:
                matches.append((keyword, line, match_ratio))
    return matches

# Find matches
matches = find_fuzzy_matches(log_data, keywords)

# Print the matches
for match in matches:
    print(f"Keyword: '{match[0]}' found in log line: '{match[1]}' with match ratio: {match[2]:.2f}")
```

Reply:
yes I know about that fact. and as I am usually the one with the latest docker engine I am not so scared. Also the deployments are usually far behind.

Reply:
did not consider any fuzzy thing.